internal/app/machined/pkg/controllers/runtime/loaded_kernel_module.go

```diff
@@ -0,0 +1,195 @@
+			}
+
+		case err = <-notifyErrCh:
+			return fmt.Errorf("error watching /proc/modules: %w", err)
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		r.ResetRestartBackoff()
+	}
+}
+
+// Module represents a kernel module parsed from /proc/modules.
+type Module struct {
+	Name         string
+	Size         int
+	Instances    int
+	Dependencies []string
+	State        string
+	Address      string
+}
+
+// ParseModules parses the contents of /proc/modules from the given reader
+// and returns a slice of module structs representing each loaded kernel module.
+func ParseModules(r io.Reader) ([]Module, error) {
+	var modules []Module
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+		if len(fields) < 6 {
+			continue // malformed line
+		}
+
+		name := fields[0]
+
+		size, err := strconv.Atoi(fields[1])
+		if err != nil {
+			return nil, fmt.Errorf("invalid size for module %s: %v", name, err)
+		}
+
+		instances, err := strconv.Atoi(fields[2])
+		if err != nil {
+			return nil, fmt.Errorf("invalid instance count for module %s: %v", name, err)
+		}
+
+		deps := []string{}
+		if fields[3] != "-" {
+			deps = slices.DeleteFunc(
+				strings.Split(fields[3], ","),
+				func(s string) bool { return s == "" },
+			)
+		}
+
+		modules = append(modules, Module{
+			Name:         name,
+			Size:         size,
+			Dependencies: deps,
+			Instances:    instances,
+			State:        fields[4],
+			Address:      fields[5],
+		})
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error scanning modules: %w", err)
+	}
+
+	return modules, nil
+}
+
+func (ctrl *LoadedKernelModuleController) reconcile(ctx context.Context, r controller.Runtime, path string) error {
+	f, err := os.OpenFile(path, os.O_RDONLY, 0)
+	if err != nil {
+		return fmt.Errorf("error opening %s: %w", path, err)
+	}
+
+	rawModules, err := ParseModules(f)
+	if err != nil {
+		return fmt.Errorf("error parsing modules from %s: %w", path, err)
+	}
+
+	// create a map to track which modules were touched
+	touchedIDs := make(map[string]struct{}, len(rawModules))
+
+	for _, module := range rawModules {
+		if err := safe.WriterModify(ctx, r,
+			runtime.NewLoadedKernelModule(runtime.NamespaceName, module.Name),
+			func(res *runtime.LoadedKernelModule) error {
+				res.TypedSpec().Name = module.Name
+				res.TypedSpec().Size = module.Size
+				res.TypedSpec().Instances = module.Instances
+				res.TypedSpec().Dependencies = module.Dependencies
+				res.TypedSpec().State = module.State
+				res.TypedSpec().Address = module.Address
+
+				return nil
+			},
+		); err != nil {
+			return fmt.Errorf("error updating LoadedKernelModule resource: %w", err)
+		}
+
+		// mark the module as touched
+		touchedIDs[module.Name] = struct{}{}
+	}
+
+	// remove kernel modules that were not touched
+	modules, err := safe.ReaderListAll[*runtime.LoadedKernelModule](ctx, r)
```

Comment:
we can simplify this a little bit, but might be easier to do as a pair programming (this is specific to new way to manage outputs in controller)